### PR TITLE
feat: inform user for lack of connection

### LINKED ISSUE #211: Inform user for lack of connection to the server
### Description

Let the user know when there is no connection to the server in the Search View

Tasks
* [ ] Display a message that informs the user there is no Connection
* [ ] Hide the default message/search button
* [ ] Try reconnecting in the background

### Screenshots

_No response_

### Additional information

_No response_

### 🥦 Browser

Mozilla Firefox

### 👀 Have you checked if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Contributing Guidelines?

- [X] I have read and understood the rules in the [Contributing Guidelines](https://github.com/Dun-sin/Whisper/blob/main/CONTRIBUTING.md)

### Are you willing to work on this issue ?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by mathiasayivor**
# Fixes Issue

closes #211

# 👨‍💻 Changes proposed(What did you do ?)
* Inform user when connection to server is lost

# ✔️ Check List (Check all the applicable boxes)

- [x] My code follows the code style of this project.
- [x] This PR does not contain plagiarized content.
- [x] The title and description of the PR is clear and explains the approach.

##  Note to reviewers
# 📷 Screenshots

### CODE REVIEW

File: client/src/components/BuddyMatcher.jsx [hunk 1]
```diff
@@ -28,7 +32,10 @@ const BuddyMatcher = () => {
     const startNewSearch = () => {
         startSearch();
         setLoadingText(defaultLoadingText);
-        socket.emit('join', { loginId: authState.loginId, email: authState.email });
+        socket.volatile.emit('join', {
+            loginId: authState.loginId,
+            email: authState.email,
+        });
```
Comment (Dun-sin) [inline client/src/components/BuddyMatcher.jsx:38]: what's the reason for this change here?

File: client/src/components/NavBar.jsx [hunk 2]
```diff
@@ -62,7 +62,7 @@ const NavBar = ({ className }) => {
     const fullscreenPages = ['/founduser'];
 
     const hideNavbar = useMemo(
-        () => fullscreenPages.includes(location.pathname) && !app.isSearching,
+        () => fullscreenPages.includes(location.pathname) && app.currentChatId,
```
Comment (Dun-sin) [inline client/src/components/NavBar.jsx:65]: what's the reason of this change here?

File: client/src/components/BuddyMatcher.jsx (hunk 1 above)
Comment (mathiasayivor) [reply client/src/components/BuddyMatcher.jsx:38]: To avoid sending that event multiple times to the server once the connection is back

File: client/src/components/NavBar.jsx (hunk 2 above)
Comment (mathiasayivor) [reply client/src/components/NavBar.jsx:65]: Only show the navbar when we're not currently chatting.

The previous one works, but only when searching for a new buddy. The new change makes it available for the alert screen as well

Comment (Dun-sin) [reply client/src/components/NavBar.jsx:65]: please remove that and create a new issue for it with it's solution

File: client/src/components/BuddyMatcher.jsx (hunk 1 above)
Comment (Dun-sin) [reply client/src/components/BuddyMatcher.jsx:38]: sounds good, i noticed it's only this emit that uses it and not the other login emit, why?

File: client/src/components/NavBar.jsx (hunk 2 above)
Comment (mathiasayivor) [reply client/src/components/NavBar.jsx:65]: It's also part of the feature imo. Is there a special reason for creating a whole new issue for it?

File: client/src/components/BuddyMatcher.jsx (hunk 1 above)
Comment (mathiasayivor) [reply client/src/components/BuddyMatcher.jsx:38]: Which login?

File: client/src/components/NavBar.jsx (hunk 2 above)
Comment (Dun-sin) [reply client/src/components/NavBar.jsx:65]: It's not, please create a new issue

File: client/src/components/BuddyMatcher.jsx (hunk 1 above)
Comment (Dun-sin) [reply client/src/components/BuddyMatcher.jsx:38]: Search for "socket. Emit(" login")" on vs code, all other results in the client is what I mean

File: client/src/components/NavBar.jsx (hunk 2 above)
Comment (mathiasayivor) [reply client/src/components/NavBar.jsx:65]: How is it not? Can you try removing that line to see why it's the same feature?

Comment (Dun-sin) [reply client/src/components/NavBar.jsx:65]: I see, my mistake for the misunderstanding

File: client/src/components/BuddyMatcher.jsx (hunk 1 above)
Comment (mathiasayivor) [reply client/src/components/BuddyMatcher.jsx:38]: I can't find any results for `socket.emit('login')`, the closest I can find is `socket.emit('logout')` is there a particular file or line I can check?

Comment (mathiasayivor) [reply client/src/components/BuddyMatcher.jsx:38]: Oh, I've just realised there's already a `join` event being emitted once the socket connects. I just need to connect. Plus I need make all the `join` events volatile.

I guess you were referring to the second `join` event at line 128-131 (this same file)?

Comment (Dun-sin) [reply client/src/components/BuddyMatcher.jsx:38]: Yes and every instance of a join socket event, thought there was a specific reason why you didn't change that too

File: client/src/components/BuddyMatcher.jsx [hunk 3]
```diff
@@ -51,9 +51,9 @@ const BuddyMatcher = () => {
             return;
         }
 
+        startSearch()
+        setLoadingText(defaultLoadingText)
```
Comment (Dun-sin) [inline client/src/components/BuddyMatcher.jsx:55]: what's with this one?

Comment (mathiasayivor) [reply client/src/components/BuddyMatcher.jsx:55]: The first code was calling `socket.connect()` and `startNewSearch()` as a result, the `join` event is emitted twice;
* First, emit is immediately we call `startNewSearch()`:
https://github.com/Dun-sin/Whisper/blob/ae221c811648c747e5818546f7fff71850247e9e/client/src/components/BuddyMatcher.jsx#L32
* Second emit after the socket connects:
https://github.com/Dun-sin/Whisper/blob/ae221c811648c747e5818546f7fff71850247e9e/client/src/components/BuddyMatcher.jsx#L112

Line 54-55 is to restart the loading animation.

Comment (mathiasayivor) [reply client/src/components/BuddyMatcher.jsx:55]: It took me long enough to figure out my comment was not submitted. It was `PENDING`

Comment (Dun-sin) [reply client/src/components/BuddyMatcher.jsx:55]: I get it, thanks for explaining

**Review by Dun-sin — APPROVED** [review]
Looks good, thanks
